adb/fastboot.py

```diff
@@ -310,20 +321,19 @@ def Download(self, source_file, source_len=0,
         Returns:
           Response to a download request, normally nothing.
         """
+
         if isinstance(source_file, str):
+            source_file_path = str(source_file)
             source_len = os.stat(source_file).st_size
-            source_file = open(source_file)
+            with open(source_file_path, 'rb') as fh:
+                source_file = BytesIO(fh.read())
```

Comment:
do we still need the data.encode('utf8') or was that a bug in the first place?

Reply:
By data.encode('utf-8') do you mean the BytesIO creation?

The behavior here was improved so that a caller to Download() can either pass in a file-like object directly or a string file path.  In the event they pass in a string, we have to convert it into a file like object before it's passed into HandleDataSending().  This statement was previously leaving a file handle open, so we read the contents and put it into BytesIO so that it has file-like properties.

If this isn't what you meant then let me know!

Reply:
Oh I see what you're referring too.  The diff below shows us running data.encode('utf-8') on the file contents.  This was a bug because the images that are Downloaded are almost always binary/outside the utf-8 space and would fail.